src/api/utils/tradeQuote/tradequote.ts

```diff
@@ -0,0 +1,404 @@
+/*
+  Copyright 2021 Set Labs Inc.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+'use strict';
+
+import { BigNumber, FixedNumber, utils as ethersUtils } from 'ethers';
+import BigDecimal from 'js-big-decimal';
+
+import {
+  CoinGeckoCoinPrices,
+  CoinGeckoTokenMap,
+  QuoteOptions,
+  TradeQuote,
+  TokenResponse,
+} from '../../../types/index';
+
+import SetTokenAPI from '../../SetTokenAPI';
+
+import {
+  CoinGeckoDataService,
+  USD_CURRENCY_CODE
+} from './coingecko';
+
+import { GasOracleService } from './gasOracle';
+import { ZeroExTradeQuoter } from './zeroex';
+
+export const ZERO_EX_ADAPTER_NAME = 'ZeroExApiAdapterV3';
+import { Address } from '@setprotocol/set-protocol-v2/utils/types';
+
+const SCALE = BigNumber.from(10).pow(18);
+
+/**
+ * @title TradeQuoteAPI
+ * @author Set Protocol
+ *
+ * A utility library to generate trade quotes for token pairs associated with a
+ * set for Ethereum and Polygon chains. Uses 0xAPI to get quote and requires a valid
+ * 0x api key.
+ */
+
+export class TradeQuoteAPI {
+  private setToken: SetTokenAPI;
+  private tokenMap: CoinGeckoTokenMap;
+  private largeTradeGasCostBase: number = 150000;
+  private tradeQuoteGasBuffer: number = 5;
+  private zeroExApiKey: string;
+
+  constructor(setToken: SetTokenAPI, zeroExApiKey: string = '') {
+    this.setToken = setToken;
+    this.zeroExApiKey = zeroExApiKey;
+  }
+
+  /**
+   * Generates a trade quote for a token pair in a SetToken. This method requires
+   * a token metadata map (passed with the options) which can be generated using
+   * the CoinGeckoDataService in '.api/utils/coingecko.ts'.
+   *
+   * @param  options  QuoteOptions: options / config to generate the quote
+   * @return          TradeQuote: trade quote object
+   */
+  async generate(options: QuoteOptions): Promise<TradeQuote> {
+    this.tokenMap = options.tokenMap;
+    const feePercentage = options.feePercentage || 0;
+    const isFirmQuote = options.isFirmQuote || false;
+    const chainId = options.chainId;
+    const exchangeAdapterName = ZERO_EX_ADAPTER_NAME;
```

Comment:
Validate rest of `QuoteOptions`

Reply:
Required option validations are now happening in [TradeAPI#fetchTradeQuote][1]

[1]: https://github.com/SetProtocol/set.js/blob/2c360f0f140a512e11c0ad0ddc9c6de6c9bc2caf/src/api/TradeAPI.ts#L172-L177